 ### Sentiment analysis on skin care products

##  1. Download data from [Amazon product data](https://jmcauley.ucsd.edu/data/amazon/)

*We need "reviews_Beauty.json.gz" and "meta_Beauty.json.gz" files.*

##  2. Read the data into pandas data frame

#### 2.1 Reading the review dataset into a data frame

In [43]:
import pandas as pd
import gzip
import json

def parse(path):
    g = gzip.open(path, 'rb')
    for l in g:
        yield json.loads(l)

def getDF(path):
    i = 0
    df = {}
    for d in parse(path):
        df[i] = d
        i += 1
    return pd.DataFrame.from_dict(df, orient='index')

df_reviews = getDF("reviews_Beauty.json.gz")

In [44]:
df_reviews

reviewerID        asin    reviewerName helpful  \
0        A39HTATAQ9V7YF  0205616461  cheryl roberts  [0, 0]   
1        A3JM6GV9MNOF9X  0558925278           Patty  [0, 1]   
2        A1Z513UWSAAO0F  0558925278        Renita M  [0, 0]   
3        A1WMRR494NWEWV  0733001998  Amazon Shopper  [0, 0]   
4        A3IAAVS479H7M7  0737104473       MAC Lover  [2, 2]   
...                 ...         ...             ...     ...   
2023065  A3DEHKPFANB8VA  B00LORWRJA    Susan Whalen  [0, 0]   
2023066  A3DEHKPFANB8VA  B00LOS7MEE    Susan Whalen  [0, 0]   
2023067   AG9TJLJUN5OM3  B00LP2YB8E           Karen  [0, 0]   
2023068   AYBIB14QOI9PC  B00LPVG6V0    ellen floren  [0, 0]   
2023069  A1PB6OT0OOPKNQ  B00LU0LTOU        David H.  [1, 1]   

                                                reviewText  overall  \
0        i do love this moisturizer and would recommend...      5.0   
1        I received this product before the deadline.I ...      3.0   
2        I love this set. Great buy for the price. I do...      5.0   
3        A nice moisturizer, all natural ingredients an...      4.0   
4        Please research the MAC Hello Kitty collection...      1.0   
...                                                    ...      ...   
2023065  Beautiful gold polish! The color is beautiful,...      5.0   
2023066  Worth every penny! All I can say is that my na...      5.0   
2023067                          Beautiful, fast shipping.      5.0   
2023068  I stared using this wash last fall as I always...      5.0   
2023069  This ECOCRAFTWORLD bag is so sturdy and is gre...      5.0   

                                                   summary  unixReviewTime  \
0                              bio-active anti-aging serum      1369699200   
1        This product is ok, I'm use Baby Kabuki in moment      1355443200   
2                                          I love this set      1404691200   
3                                         Nice Moisturizer      1382572800   
4                                                 Fake MAC      1274227200   
...                                                    ...             ...   
2023065                          Love the color and value!      1405296000   
2023066                           Pretty and long lasting.      1405296000   
2023067                                         Five Stars      1405382400   
2023068  I stared using this wash last fall as I always...      1405555200   
2023069  This ECOCRAFTWORLD bag is so sturdy and is gre...      1405641600   

          reviewTime  
0        05 28, 2013  
1        12 14, 2012  
2         07 7, 2014  
3        10 24, 2013  
4        05 19, 2010  
...              ...  
2023065  07 14, 2014  
2023066  07 14, 2014  
2023067  07 15, 2014  
2023068  07 17, 2014  
2023069  07 18, 2014  

[2023070 rows x 9 columns]

*The data has 2023070 entries and 9 columns:*

- reviewerID - ID of the reviewer
- asin - ID of the product
- reviewerName - name of the reviewer 
- helpful - helpfulness rating of the review
- reviewText - text of the review 
- overall - rating of the product 
- summary - summary of the review 
- unixReviewTime - time of the review (unix time) 
- reviewTime - time of the review (raw)

#### 2.2 Reading the meta data into a data frame

In [102]:
def parseMeta(path): 
    g = gzip.open(path, 'r')
    for l in g:
        yield json.dumps(eval(l))

f = open("output.strict", 'w')
for l in parseMeta("meta_Beauty.json.gz"):
    f.write(l + '\n')

In [103]:
def getDF_meta(path):
    df = pd.read_json(path, lines=True)
    return df

df_meta = getDF_meta("output.strict")

In [104]:
df_meta

asin                                        description  \
0       0205616461  As we age, our once youthful, healthy skin suc...   
1       0558925278  Mineral Powder Brush--Apply powder or mineral ...   
2       0733001998  From the Greek island of Chios, this Mastiha b...   
3       0737104473  Limited edition Hello Kitty Lipstick featuring...   
4       0762451459  The mermaid is an elusive (okay, mythical) cre...   
...            ...                                                ...   
259190  B00LJEACWC  Would you like that perfect natural makeup loo...   
259191  B00LJPYR8G  "In the garden of the senses lies a pathway to...   
259192  B00LKNVKTQ  The Babyliss MiraCurl Nano Titanium Curl Machi...   
259193  B00LLPT4HI  Ingredients:\n100% Pure, Unrefined, Organic Af...   
259194  B00LMOT7X4  Dare to wear Maybelline New York Eye Studio Co...   

                                                    title  \
0       Bio-Active Anti-Aging Serum (Firming Ultra-Hyd...   
1       Eco Friendly Ecotools Quality Natural Bamboo C...   
2                                     Mastiha Body Lotion   
3       Hello Kitty Lustre Lipstick (See sellers comme...   
4             Stephanie Johnson Mermaid Round Snap Mirror   
...                                                   ...   
259190  Kabuki brush By Simply Essentials. Today 50% o...   
259191  Kuumba Made Fragrances (Tunisian Myrrh, 1/4oz ...   
259192              Babyliss Pro - Nano Titanium Miracurl   
259193  100% Organic Raw Unrefined African Shea Butter...   
259194  Maybelline Color Tattoo Leather 24HR Cream Gel...   

                                                    imUrl  \
0       http://ecx.images-amazon.com/images/I/41DecrGO...   
1       http://ecx.images-amazon.com/images/I/51L%2BzY...   
2       http://ecx.images-amazon.com/images/I/311WK5y1...   
3       http://ecx.images-amazon.com/images/I/31u6Hrzk...   
4       http://ecx.images-amazon.com/images/I/41y2%2BF...   
...                                                   ...   
259190  http://ecx.images-amazon.com/images/I/51kNB8qr...   
259191  http://ecx.images-amazon.com/images/I/21XbzJEx...   
259192  http://ecx.images-amazon.com/images/I/4138wRfV...   
259193  http://ecx.images-amazon.com/images/I/41up5%2B...   
259194  http://ecx.images-amazon.com/images/I/4110Hzm0...   

                                 salesRank  \
0       {'Health & Personal Care': 461765}   
1                       {'Beauty': 402875}   
2                       {'Beauty': 540255}   
3                       {'Beauty': 931125}   
4                                      NaN   
...                                    ...   
259190                   {'Beauty': 16197}   
259191                  {'Beauty': 301257}   
259192                     {'Beauty': 727}   
259193                    {'Beauty': 1265}   
259194                  {'Beauty': 141114}   

                                               categories   price  \
0       [[Beauty, Skin Care, Face, Creams & Moisturize...     NaN   
1       [[Beauty, Tools & Accessories, Makeup Brushes ...     NaN   
2       [[Beauty, Skin Care, Body, Moisturizers, Lotio...     NaN   
3                      [[Beauty, Makeup, Lips, Lipstick]]     NaN   
4       [[Beauty, Tools & Accessories, Mirrors, Makeup...   19.98   
...                                                   ...     ...   
259190  [[Beauty, Tools & Accessories, Makeup Brushes ...   16.99   
259191     [[Beauty, Fragrance, Women's, Essential Oils]]   14.99   
259192  [[Beauty, Hair Care, Styling Tools, Irons, Cur...  139.00   
259193  [[Beauty, Skin Care, Body, Moisturizers, Lotio...    9.97   
259194               [[Beauty, Makeup, Eyes, Eye Shadow]]   14.72   

                                                  related       brand  
0                                                     NaN         NaN  
1                                                     NaN         NaN  
2                                                     NaN         NaN  
3

*Meta data includes product description, title, urls, salesRank, categories, price, related products and brand.*



##  3. Create skin care dataset

#### 3.1 Choosing skin care products

*Since we only interested in 'Skin Care' products, we select the products that includes 'Skin Care' in their categories.*   

In [69]:
selected_ind = []
for i in range(len(df_meta)):
    if 'Skin Care' in df_meta.categories[i][0]:
        selected_ind.append(i)

In [73]:
len(selected_ind)

64380

In [76]:
df_skin_care = df_meta.iloc[selected_ind]

In [77]:
df_skin_care

asin                                        description  \
0       0205616461  As we age, our once youthful, healthy skin suc...   
2       0733001998  From the Greek island of Chios, this Mastiha b...   
9       1304146537  Eyliplex-2 is a dual solution that focuses on ...   
10      130414643X  Eyliplex-2 is a dual solution that focuses on ...   
11      1304174905  INSTANTLY REVITALIZE SKIN MOIST THE FACE OF HI...   
...            ...                                                ...   
259182  B00LH28Q88  No preservatives, chemicals, or added fragranc...   
259186  B00LIFR8SS  Composed of plant extracts and essential oils,...   
259187  B00LIF47O6  The Skin Shield with essential oils is a calmi...   
259188  B00LIBEPDS  The small Icy Cool Roller makes it easy to man...   
259193  B00LLPT4HI  Ingredients:\n100% Pure, Unrefined, Organic Af...   

                                                    title  \
0       Bio-Active Anti-Aging Serum (Firming Ultra-Hyd...   
2                                     Mastiha Body Lotion   
9       Set of 2 Goodskin Labs Eyliplex-2 Eye Life and...   
10      Goodskin Labs Eyliplex-2 Eye Life and Circle R...   
11      *Macely* Good Afternoon Rose Lemon Tea Bb 30ml...   
...                                                   ...   
259182  Argan Oil for Hair, Face, Skin and Nails! 100%...   
259186  La Formule - Purifying Skin Wash - Soap-free c...   
259187  La Formule - Skin Shield for Problem Skin with...   
259188                                    Icy cool roller   
259193  100% Organic Raw Unrefined African Shea Butter...   

                                                    imUrl  \
0       http://ecx.images-amazon.com/images/I/41DecrGO...   
2       http://ecx.images-amazon.com/images/I/311WK5y1...   
9       http://ecx.images-amazon.com/images/I/51vx4ZM1...   
10      http://ecx.images-amazon.com/images/I/41Bps9Pm...   
11      http://ecx.images-amazon.com/images/I/416SYK3g...   
...                                                   ...   
259182  http://ecx.images-amazon.com/images/I/41lmOrBq...   
259186  http://ecx.images-amazon.com/images/I/314HcyXj...   
259187  http://ecx.images-amazon.com/images/I/31ptUNyE...   
259188  http://ecx.images-amazon.com/images/I/31L9zfel...   
259193  http://ecx.images-amazon.com/images/I/41up5%2B...   

                                 salesRank  \
0       {'Health & Personal Care': 461765}   
2                       {'Beauty': 540255}   
9                       {'Beauty': 371552}   
10                      {'Beauty': 257530}   
11                                     NaN   
...                                    ...   
259182  {'Health & Personal Care': 359729}   
259186                  {'Beauty': 381959}   
259187                  {'Beauty': 278272}   
259188                   {'Beauty': 14379}   
259193                    {'Beauty': 1265}   

                                               categories  price  \
0       [[Beauty, Skin Care, Face, Creams & Moisturize...    NaN   
2       [[Beauty, Skin Care, Body, Moisturizers, Lotio...    NaN   
9       [[Beauty, Skin Care, Eyes, Dark Circle Treatme...    NaN   
10      [[Beauty, Skin Care, Eyes, Dark Circle Treatme...    NaN   
11      [[Beauty, Skin Care, Face, Creams & Moisturize...   9.99   
...                                                   ...    ...   
259182  [[Beauty, Skin Care, Face, Creams & Moisturize...  16.95   
259186     [[Beauty, Skin Care, Face, Cleansers, Washes]]  26.39   
259187             [[Beauty, Skin Care, Face, Cleansers]]  26.98   
259188  [[Beauty, Skin Care, Eyes, Dark Circle Treatme...  29.95   
259193  [[Beauty, Skin Care, Body, Moisturizers, Lotio...   9.97   

                                                  related brand  
0                                                     NaN   NaN  
2                                                     NaN   NaN  
9       {'also_bought': ['B002J2KOXK', 'B008TCVZ6Y'], ...   NaN  
10      {'also_bought': ['B002J2KOXK'

*Now we have left with 64380 skin care products.* 

#### 3.2 Choosing skin care reviews

*We select skin care reviews from our review data frame. We select reviews if their asin (product id) is in our selected skin care products list.* 

In [78]:
asin_skin_care = list(df_skin_care.asin)

In [81]:
reviews_skin_care = df.loc[df['asin'].isin(asin_skin_care)]


In [82]:
reviews_skin_care

reviewerID        asin  \
0        A39HTATAQ9V7YF  0205616461   
3        A1WMRR494NWEWV  0733001998   
9        A327B0I7CYTEJC  130414643X   
10       A1BG8QW55XHN6U  130414643X   
11        AIFAAVTUYHEHB  130414643X   
...                 ...         ...   
2023042  A3LGT6UZL99IW1  B00LLPT4HI   
2023043   ASKZO80Z1RKTR  B00LLPT4HI   
2023044  A1A7EBY8XJ46PA  B00LLPT4HI   
2023045  A1HQ0KTGWQIWYI  B00LLPT4HI   
2023046  A3UJRNI8UR4871  B00LLPT4HI   

                                             reviewerName helpful  \
0                                          cheryl roberts  [0, 0]   
3                                          Amazon Shopper  [0, 0]   
9                                            Sheri Hamlin  [0, 0]   
10                                               Victoria  [0, 0]   
11                                                   zoya  [0, 0]   
...                                                   ...     ...   
2023042  Richard C. Drew "Anaal Nathra/Uthe vas Bethod...  [0, 0]   
2023043                                         Robin Lee  [1, 1]   
2023044                                            R.Wang  [0, 0]   
2023045                                      Sunny SMOOTH  [0, 0]   
2023046                               Wulfstan "wulfstan"  [0, 1]   

                                                reviewText  overall  \
0        i do love this moisturizer and would recommend...      5.0   
3        A nice moisturizer, all natural ingredients an...      4.0   
9        Dark circles under the eyes runs in my family ...      4.0   
10       I have purchased on Amazon . This was the firs...      5.0   
11       Is there  any difference between the one typed...      4.0   
...                                                    ...      ...   
2023042  I have very dry elbows and knees.  I have a to...      5.0   
2023043  This 100% Organic Raw Unrefined African Shea B...      5.0   
2023044  I live in a hot dry climate and didn't find th...      5.0   
2023045  Delivery time was very fast. This butter works...      5.0   
2023046  This is 100% pure Shea Butter. Do not mistake ...      5.0   

                                              summary  unixReviewTime  \
0                         bio-active anti-aging serum      1369699200   
3                                    Nice Moisturizer      1382572800   
9                           I really like this stuff.      1389052800   
10                                  love this product      1372032000   
11                  the smell and color are different      1378252800   
...                                               ...             ...   
2023042                        This stuff is amazing!      1405382400   
2023043  "Wonderful and Soothing Moisturizer"........      1405209600   
2023044                                     TERRIFIC!      1405728000   
2023045                Fast delivery, and smooth skin      1405900800   
2023046                             The "Real Stuff"!      1405987200   

          reviewTime  
0        05 28, 2013  
3        10 24, 2013  
9         01 7, 2014  
10       06 24, 2013  
11        09 4, 2013  
...              ...  
2023042  07 15, 2014  
2023043  07 13, 2014  
2023044  07 19, 2014  
2023045  07 21, 2014  
2023046  07 22, 2014  

[629299 rows x 9 columns]

*Now let's save the data frames into csv files.*

In [89]:
reviews_skin_care.to_csv(r'reviews_skin_care.csv.gz', index=False, compression='gzip')

In [90]:
df_skin_care.to_csv(r'meta_skin_care.csv.gz', index=False, compression='gzip')